src/main/java/net/glowstone/generator/decorators/overworld/MushroomDecorator.java

```diff
@@ -51,18 +56,30 @@ public void decorate(World world, Random random, Chunk source) {
 
                 Block block = world.getBlockAt(x, y, z);
                 Block blockBelow = world.getBlockAt(x, y - 1, z);
-                if (y < 255 && block.getType() == Material.AIR &&
-                    ((blockBelow.getType() == Material.GRASS
-                        || blockBelow.getState().getData() instanceof Dirt &&
-                        ((Dirt) blockBelow.getState().getData()).getType() != DirtType.PODZOL)
-                        && block.getLightLevel() < 13 ||
-                        blockBelow.getType() == Material.MYCEL
-                        || blockBelow.getState().getData() instanceof Dirt &&
-                        ((Dirt) blockBelow.getState().getData()).getType() == DirtType.PODZOL)) {
-                    BlockState state = block.getState();
-                    state.setType(type);
-                    state.setData(new MaterialData(type));
-                    state.update(true);
+                if (y < 255 && block.getType() == Material.AIR) {
+                    boolean canPlaceShroom;
+                    switch (blockBelow.getType()) {
+                        case MYCEL:
+                            canPlaceShroom = true;
+                            break;
+                        case GRASS:
+                            canPlaceShroom = (block.getLightLevel() < 13);
+                            break;
+                        default:
+                            MaterialData data = blockBelow.getState().getData();
+                            if (data instanceof Dirt) {
```

Comment:
add a material case for DIRT and then do these additional checks, rather than in default

Reply:
Done.